lib/Sema/TypeCheckAvailability.cpp

```diff
@@ -3023,6 +3023,88 @@ static bool diagnoseIsolatedAnyAvailability(
       ReferenceDC);
 }
 
+/// Make sure the generic arguments conform to all known invertible protocols.
+/// Runtimes prior to NoncopyableGenerics do not check if any of the
+/// generic arguments conform to Copyable/Escapable during dynamic casts.
+/// But a dynamic cast *needs* to check if the generic arguments conform,
+/// to determine if the cast should be permitted at all. For example:
+///
+///    struct X<T> {}
+///    extension X: P where T: Y {}
+///
+///     func f<Y: ~Copyable>(...) {
+///       let x: X<Y> = ...
+///       _ = x as? any P   // <- cast should fail
+///     }
+///
+/// The dynamic cast here must fail because Y does not conform to Copyable,
+/// thus X<Y> doesn't conform to P!
+///
+/// \param boundTy The generic type with its generic arguments.
+/// \returns the invertible protocol for which a conformance is missing in
+///          one of the generic arguments, or none if all are present for
+///          every generic argument.
+static std::optional<InvertibleProtocolKind> checkGenericArgsForInvertibleReqs(
+    BoundGenericType *boundTy) {
+  auto *decl = boundTy->getDecl();
+  for (auto arg : boundTy->getGenericArgs()) {
+
+    if (arg->hasTypeParameter() || arg->hasUnboundGenericType())
+      arg = decl->mapTypeIntoContext(arg);
```

Comment:
- How do you know `arg` is a valid type parameter for the generic signature of the decl? It won't be in general, but this isn't covered by your tests.
- Also it doesn't appear that the hasUnboundTypeParameter() case isn't covered by your tests.
- Since this is called with the Type of an Expr, there shouldn't be any type parameters in there ever, and no unbound generic types after solution application, so this entire `if` statement might be dead.

Reply:
Thanks!